plugins/mission/mission_impl.cpp

```diff
@@ -882,4 +884,107 @@ void MissionImpl::process_timeout()
+    if (!file) { // File open error
+        return Mission::Result::FAILED_TO_OPEN_QGC_PLAN;
+    }
+
+    // Read QGC plan into a string stream
+    std::stringstream ss;
+    ss << file.rdbuf();
+    file.close();
+
+    std::string err;
+    const auto parsed_plan = Json::parse(ss.str(), err); // parse QGC plan
+    if (!err.empty()) { // Parse error
+        return Mission::Result::FAILED_TO_PARSE_QGC_PLAN;
+    }
+
+    // Compose mission items
+    return compose_mission_items_from_json(mission_items, parsed_plan);
+}
+
+Mission::Result
+MissionImpl::compose_mission_items_from_json(Mission::mission_items_t &mission_items,
+                                             const Json &qgc_plan_json)
+{
+    const auto json_mission_items = qgc_plan_json["mission"];
+
+    // Compose mission items by iterating JSON mission items
+    for (auto &json_mission_item : json_mission_items["items"].array_items()) {
+        // Parameters of Mission item & MAV command of it.
+        float speed_m_s = 0.f;
+        float gimbal_pitch_deg = 0.f, gimbal_yaw_deg = 0.f;
+        bool is_fly_through = false;
+        double lat_deg = NAN, lon_deg = NAN, rel_alt_deg = NAN;
+        double photo_inteval = 0.;
+        auto camera_action = MissionItem::CameraAction::NONE;
+        auto command = json_mission_item["command"].int_value();
+
+        // Extract parameters of each mission item
+        std::vector<double> params;
+        for (auto &p : json_mission_item["params"].array_items()) {
+            params.push_back(p.number_value());
+        }
+
+        switch (command) {
+            case MAV_CMD_IMAGE_START_CAPTURE:
+                photo_inteval = params[1];
+                camera_action = MissionItem::CameraAction::START_PHOTO_INTERVAL;
+                break;
+            case MAV_CMD_VIDEO_START_STREAMING:
+                camera_action = MissionItem::CameraAction::START_VIDEO;
+                break;
+            case MAV_CMD_VIDEO_STOP_STREAMING:
+                camera_action = MissionItem::CameraAction::STOP_VIDEO;
+                break;
+            case MAV_CMD_DO_CHANGE_SPEED:
+                speed_m_s = params[1];
+                break;
+            case MAV_CMD_DO_MOUNT_CONTROL:
+                gimbal_pitch_deg = -params[0]; // Pitch value is -ve as its in NED frame.
+                gimbal_yaw_deg = params[2];
+                break;
+            case MAV_CMD_NAV_WAYPOINT:
+                is_fly_through = (params[0] == 0.0) ? true : false;
+            case MAV_CMD_NAV_TAKEOFF:
+                lat_deg = params[4];
+                lon_deg = params[5];
+                rel_alt_deg = params[6];
+                break;
+            case MAV_CMD_NAV_RETURN_TO_LAUNCH:
+                break;
+            default: // Unsupported mission command
+                LogWarn() << "Mission: Unsupported mission command: " << command << ". Ignoring.";
+                break;
+        }
+        // Maybe removed
```

Comment:
Maybe what? 😄

Reply:
I meant that we can remove debug log if its annoying for larger mission. :)

Reply:
I understand but it's a funny comment.